robot_smach_states/src/robot_smach_states/navigation/topological_navigation.py

```diff
@@ -0,0 +1,162 @@
+    if msg.destination.area:
+        destination_entity_designator = EntityByIdDesignator(robot, msg.destination.entity)
+        look_at_designator = EntityByIdDesignator(robot, msg.destination.entity)
+        return NavigateToSymbolic(
+            robot=robot,
+            entity_designator_area_name_map={destination_entity_designator: msg.destination.area},
+            entity_lookat_designator=look_at_designator,
+        )
+    else:  # Assume the robot has to navigate to a waypoint
+        waypoint_designator = EntityByIdDesignator(robot, msg.destination.entity)
+        return NavigateToWaypoint(
+            robot=robot,
+            waypoint_designator=waypoint_designator,
+        )
+
+
+def convert_open_door_msg_to_action(robot: Robot, msg: Edge) -> OpenDoor:
+    door_designator = EntityByIdDesignator(robot, msg.destination.entity)
+
+    return OpenDoor(robot=robot, door_designator=door_designator)
+
+
+# ToDo: replace userdata by designators
+@smach.cb_interface(outcomes=["unreachable", "goal_not_defined", "goal_ok", "preempted"], output_keys=["action_plan"])
+def get_topological_action_plan(
+    userdata: smach.UserData, robot: Robot, entity_designator: EdEntityDesignator, area_designator: Designator = None
+) -> str:
+    """
+    Gets an action plan from the topological action server
+
+    :param userdata: smach userdata
+    :param robot: robot API object
+    :param entity_designator: designator resolving to the entity to navigate to
+    :param area_designator: designator resolving to a string describing the area to navigate to
+    """
+    # Queries planner
+    entity_id = entity_designator.resolve().id
+    area = area_designator.resolve() if area_designator is not None else ""
+    action_msgs = robot.topological_planner.get_plan(entity_id=entity_id, area=area)
+    rospy.loginfo(f"Actions: {action_msgs}")
+
+    try:
+        actions = convert_msgs_to_actions(robot, action_msgs)
+    except TopologicalPlannerException as e:
+        rospy.logerr(e.message)
+        return "unreachable"
+    userdata.action_plan = actions
+    return "goal_ok"
+
+
+# ToDo: replace userdata by designators
+@smach.cb_interface(
+    outcomes=["succeeded", "arrived", "blocked", "preempted"], input_keys=["action_plan"], output_keys=["failing_edge"]
+)
+def execute_topological_plan(userdata: smach.UserData) -> str:
+    """
+    Sequentially executes all actions in the action plan provided through UserData
+
+    :param userdata: smach userdata
+    :return: outcome
+    """
+    for action in userdata.action_plan:
+        result = action.execute()  # ToDo: process result
+        if result not in ["succeeded", "done", "arrived"]:
+            rospy.loginfo("action: {} had result {}".format(action, result))
+            if result == "preempted":
+                return "preempted"
+            userdata.failing_edge = action
+            return "blocked"
+        rospy.sleep(1.0)
```

Comment:
Why this sleep @LoyVanBeek @JosjaG @PetervDooren

Reply:
If I recall it was a #Hack so that the robot would wait between opening the door and trying to find a global path through the still opening door